### PR TITLE
Block/UnBlock list of users for getting certificates. 

### PR DESCRIPTION
**Pull request by asadiqbal08**
#### What are the relevant tickets?
fixes: #1135 

#### What's this PR do?
As a bootcamp admin, I often have a list of enrolled bootcampers who shouldn't get a certificate (typically because they dropped out before the bootcamp finished).

#### How should this be manually tested?
Execute the command in following format. 
```
python manage.py manage_certificates --run={RUN} --block camper1@example.com camper2@example.com camper3@example.com
python manage.py manage_certificates --generate-batch --run={RUN}
python manage.py manage_certificates --generate --run={RUN} --user={USER}
```
Certificate should not be generated for the blocked users in ~any~ provided enrollment run. 
But after, 
```
python manage.py manage_certificates --run={RUN} --unblock camper1@example.com camper2@example.com camper3@example.com
python manage.py manage_certificates --generate-batch --run={RUN}
python manage.py manage_certificates --generate --run={RUN} --user={USER}
```
Certificate should not be generated for the users. 

~#### NOTE: Currently executing the command with option `--block [USER LIST]` will block the users for getting certificate in ~any~ specific enrollment run. We are not targeting to block user for a specific enrollment run in the command.~

### CODE REVIEW

File: klasses/admin.py [hunk 1]
```diff
@@ -45,8 +45,20 @@ class BootcampRunEnrollmentAdmin(TimestampedModelAdmin):
...
     include_created_on_in_list = True
-    list_display = ("id", "get_user_email", "bootcamp_run", "change_status", "active")
-    list_filter = ("change_status", "active", "bootcamp_run__bootcamp")
+    list_display = (
+        "id",
+        "get_user_email",
+        "bootcamp_run",
+        "change_status",
+        "active",
+        "user_certificate_is_blocked",
```
Comment (HamzaIbnFarooq) [inline klasses/admin.py:54]: kindly have a look at this comment: https://github.com/mitodl/bootcamp-ecommerce/issues/1135#issuecomment-762510159

**Review by HamzaIbnFarooq — COMMENTED** [review]
I have three questions regarding this PR (although they don't seem to be in the current scope):
1) What happens if a blocked user enrolls in a new course? according to the current implementation, certificates will start getting created for new enrollments.
2) If a certificate has already been created and then the user gets blocked, should we show that certificate to user?
3) Would there be any scenario where admin would want to block user certificates only for a specific bootcamp-run?

File: klasses/admin.py (hunk 1 above)
Comment (asadiqbal08) [reply klasses/admin.py:54]: Yes, I changed to add a mark `user` here for better understanding.

File: klasses/utils.py [hunk 2]
```diff
@@ -150,3 +155,26 @@ def unrevoke_certificate(user, bootcamp_run):
         result["msg"] = "Valid user and bootcamp_run must be provided."
 
     return result
+
+
+def manage_user_certificate_blocking(users, block_state):
+    """Block users for getting certificates in all bootcamp run enrollments"""
+    result = {"updated": False}
+    if users:
+        BootcampRunEnrollment.objects.filter(user__email__in=users, active=True).update(
```
Comment (HamzaIbnFarooq) [inline klasses/utils.py:164]:
```suggestion
        BootcampRunEnrollment.objects.filter(user__email__in=users).update(
```
might not create much impact but I guess we shouldn't leave any loophole.

**Comment by pdpinch** [comment]
Good questions @HamzaIbnFarooq 

> 1. What happens if a blocked user enrolls in a new course? according to the current implementation, certificates will start getting created for new enrollments.

The block is specific to an enrollment, not a user. So creating certificates for new enrollments is the correct behavior. 

> 2. If a certificate has already been created and then the user gets blocked, should we show that certificate to user?

We have a separate management command for revoking certificates that's designed for hiding an already created certificate. I think it makes more sense for an admin to use the revoke command, instead of adding more code logic to hide a blocked certificate.

**Comment by pdpinch** [comment]
@asadiqbal08 I would prefer that the admin user specify the run that the block applies to, instead of having it apply to all runs. Is that difficult?

> NOTE: Currently executing the command with option --block [USER LIST] will block the users for getting certificate in any enrollment. We are not targeting to block user for a specific enrollment run in the command.

**Comment by asadiqbal08** [comment]
> What happens if a blocked user enrolls in a new course? according to the current implementation, certificates will start getting created for new enrollments.


Good catch @HamzaIbnFarooq then I think adding the option of `--run` along with `--block` make sense.

File: klasses/utils.py [hunk 3]
```diff
@@ -150,3 +155,26 @@ def unrevoke_certificate(user, bootcamp_run):
...
+        state = "blocked"
+        if not block_state:
+            state = "unblocked"
+
+        result.update(
+            {
+                "updated": True,
+                "msg": "Users:{} are ".format(users)
+                + state
+                + " for getting certificate for any bootcamp enrollment.",
```
Comment (HamzaIbnFarooq) [inline klasses/utils.py:177]: we should use bootcamp title here.

File: klasses/utils.py [hunk 4]
```diff
@@ -150,3 +155,26 @@ def unrevoke_certificate(user, bootcamp_run):
         result["msg"] = "Valid user and bootcamp_run must be provided."
 
     return result
+
+
+def manage_user_certificate_blocking(users, block_state, bootcamp_run):
+    """Block users for getting certificates in all bootcamp run enrollments"""
+    result = {"updated": False}
+    if users:
+        BootcampRunEnrollment.objects.filter(
```
Comment (HamzaIbnFarooq) [inline klasses/utils.py:164]: We can improve logging for the case where no matching enrollment is found.

**Review by HamzaIbnFarooq — COMMENTED** [review]
LG, just need some extra logging (even in certificate generation for blocked users) and test cases.

File: klasses/utils.py [hunk 5]
```diff
@@ -150,3 +155,37 @@ def unrevoke_certificate(user, bootcamp_run):
...
+
+        state = "blocked"
+        if not block_state:
+            state = "unblocked"
+
+        blocked_users_emails = [run.user.email for run in run_enrollments]
+        result.update(
+            {
+                "updated": True,
+                "msg": "User(s) with email in {} ".format(blocked_users_emails)
```
Comment (HamzaIbnFarooq) [inline klasses/utils.py:186]:
```suggestion
                "msg": "User(s) with email in {} are ".format(blocked_users_emails)
```

**Review by HamzaIbnFarooq — APPROVED** [review]
LG 👍

File: klasses/utils.py (hunk 5 above)
Comment (asadiqbal08) [reply klasses/utils.py:186]: It can be a single user in list or multiple so I thought it is better to drop `is/are` from the message.

**Pull request merged**